converter.go

```diff
@@ -388,6 +387,345 @@ func arrayToString(v driver.Value, tsmode snowflakeType, params map[string]*stri
+	}
+	typOf := reflect.TypeOf(v)
+	var jsonBytes []byte
+	if tsmode == binaryType {
+		m := make(map[string]*string, valOf.Len())
+		iter := valOf.MapRange()
+		for iter.Next() {
+			val := iter.Value().Interface().([]byte)
+			if val != nil {
+				s := hex.EncodeToString(val)
+				m[stringOrIntToString(iter.Key())] = &s
+			} else {
+				m[stringOrIntToString(iter.Key())] = nil
+			}
+		}
+		jsonBytes, err = json.Marshal(m)
+		if err != nil {
+			return bindingValue{}, err
+		}
+	} else if typOf.Elem().AssignableTo(reflect.TypeOf(time.Time{})) || typOf.Elem().AssignableTo(reflect.TypeOf(sql.NullTime{})) {
+		m := make(map[string]*string, valOf.Len())
+		iter := valOf.MapRange()
+		for iter.Next() {
+			val, valid, err := toNullableTime(iter.Value().Interface())
+			if err != nil {
+				return bindingValue{}, err
+			}
+			if !valid {
+				m[stringOrIntToString(iter.Key())] = nil
+			} else {
+				typ := driverTypeToSnowflake[tsmode]
+				s, err := timeToString(val, typ, params)
+				if err != nil {
+					return bindingValue{}, err
+				}
+				m[stringOrIntToString(iter.Key())] = &s
+			}
+		}
+		jsonBytes, err = json.Marshal(m)
+		if err != nil {
+			return bindingValue{}, err
+		}
+	} else if typOf.Elem().AssignableTo(reflect.TypeOf(sql.NullString{})) {
+		m := make(map[string]*string, valOf.Len())
+		iter := valOf.MapRange()
+		for iter.Next() {
+			val := iter.Value().Interface().(sql.NullString)
+			if val.Valid {
+				m[stringOrIntToString(iter.Key())] = &val.String
+			} else {
+				m[stringOrIntToString(iter.Key())] = nil
+			}
+		}
+		jsonBytes, err = json.Marshal(m)
+		if err != nil {
+			return bindingValue{}, err
+		}
+	} else if typOf.Elem().AssignableTo(reflect.TypeOf(sql.NullByte{})) || typOf.Elem().AssignableTo(reflect.TypeOf(sql.NullInt16{})) || typOf.Elem().AssignableTo(reflect.TypeOf(sql.NullInt32{})) || typOf.Elem().AssignableTo(reflect.TypeOf(sql.NullInt64{})) {
+		m := make(map[string]*int64, valOf.Len())
+		iter := valOf.MapRange()
+		for iter.Next() {
+			val, valid := toNullableInt64(iter.Value().Interface())
+			if valid {
+				m[stringOrIntToString(iter.Key())] = &val
+			} else {
+				m[stringOrIntToString(iter.Key())] = nil
+			}
+		}
+		jsonBytes, err = json.Marshal(m)
+		if err != nil {
+			return bindingValue{}, err
+		}
+	} else if typOf.Elem().AssignableTo(reflect.TypeOf(sql.NullFloat64{})) {
+		m := make(map[string]*float64, valOf.Len())
+		iter := valOf.MapRange()
+		for iter.Next() {
+			val := iter.Value().Interface().(sql.NullFloat64)
+			if val.Valid {
+				m[stringOrIntToString(iter.Key())] = &val.Float64
+			} else {
+				m[stringOrIntToString(iter.Key())] = nil
+			}
+		}
+		jsonBytes, err = json.Marshal(m)
+		if err != nil {
+			return bindingValue{}, err
+		}
+	} else if typOf.Elem().AssignableTo(reflect.TypeOf(sql.NullBool{})) {
+		m := make(map[string]*bool, valOf.Len())
+		iter := valOf.MapRange()
+		for iter.Next() {
+			val := iter.Value().Interface().(sql.NullBool)
+			if val.Valid {
```

Comment:
lots of duplicated code, sadly not much can be done here, nor there is much sense to do it :/

Reply:
Yeah, there is no common interface :(